pytest is leaking frames of tests that fail

The test below reproduces the problem:

```python
import weakref

class Obj:
    pass

ref = None

def test1():
    obj = Obj()
    global ref
    ref = weakref.ref(obj)
    assert 0

def test2():
    assert ref() is None
```

This fails with:

```
================================== FAILURES ===================================
____________________________________ test1 ____________________________________

    def test1():
        obj = Obj()
        global ref
        ref = weakref.ref(obj)
>       assert 0
E       assert 0

test_foo.py:15: AssertionError
____________________________________ test2 ____________________________________

    def test2():
>       assert ref() is None
E       assert <test_foo.Obj object at 0x0000017302911198> is None
E        +  where <test_foo.Obj object at 0x0000017302911198> = ref()

test_foo.py:19: AssertionError
========================== 2 failed in 0.04 seconds ===========================
```

The reference to `test1`'s frame should not have leaked. If we comment the `assert 0` line, then both tests pass.

xref: pytest-dev/pytest-qt#187

--- a/_pytest/runner.py
+++ b/_pytest/runner.py
@@ -105,6 +105,7 @@ def pytest_runtest_setup(item):
 
 def pytest_runtest_call(item):
     _update_current_test_var(item, 'call')
+    sys.last_type, sys.last_value, sys.last_traceback = (None, None, None)
     try:
         item.runtest()
     except Exception:
@@ -114,7 +115,7 @@ def pytest_runtest_call(item):
         sys.last_type = type
         sys.last_value = value
         sys.last_traceback = tb
-        del tb  # Get rid of it in this namespace
+        del type, value, tb  # Get rid of these in this frame
         raise
 
 

--- a/changelog/2798.bugfix.rst
+++ b/changelog/2798.bugfix.rst
@@ -1,0 +1,3 @@
+Reset ``sys.last_type``, ``sys.last_value`` and ``sys.last_traceback`` before each test executes. Those attributes
+are added by pytest during the test run to aid debugging, but were never reset so they would create a leaking
+reference to the last failing test's frame which in turn could never be reclaimed by the garbage collector.

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -988,3 +988,33 @@ def test_fixture_order_respects_scope(testdir):
     ''')
     result = testdir.runpytest()
     assert result.ret == 0
+
+
+def test_frame_leak_on_failing_test(testdir):
+    """pytest would leak garbage referencing the frames of tests that failed that could never be reclaimed (#2798)
+
+    Unfortunately it was not possible to remove the actual circles because most of them
+    are made of traceback objects which cannot be weakly referenced. Those objects at least
+    can be eventually claimed by the garbage collector.
+    """
+    testdir.makepyfile('''
+        import gc
+        import weakref
+
+        class Obj:
+            pass
+
+        ref = None
+
+        def test1():
+            obj = Obj()
+            global ref
+            ref = weakref.ref(obj)
+            assert 0
+
+        def test2():
+            gc.collect()
+            assert ref() is None
+    ''')
+    result = testdir.runpytest_subprocess()
+    result.stdout.fnmatch_lines(['*1 failed, 1 passed in*'])

--- a/testing/test_runner.py
+++ b/testing/test_runner.py
@@ -719,24 +719,33 @@ def test_makereport_getsource_dynamic_code(testdir, monkeypatch):
     result.stdout.fnmatch_lines(["*test_fix*", "*fixture*'missing'*not found*"])
 
 
-def test_store_except_info_on_eror():
+def test_store_except_info_on_error():
     """ Test that upon test failure, the exception info is stored on
     sys.last_traceback and friends.
     """
-    # Simulate item that raises a specific exception
-    class ItemThatRaises(object):
+    # Simulate item that might raise a specific exception, depending on `raise_error` class var
+    class ItemMightRaise(object):
         nodeid = 'item_that_raises'
+        raise_error = True
 
         def runtest(self):
-            raise IndexError('TEST')
+            if self.raise_error:
+                raise IndexError('TEST')
     try:
-        runner.pytest_runtest_call(ItemThatRaises())
+        runner.pytest_runtest_call(ItemMightRaise())
     except IndexError:
         pass
     # Check that exception info is stored on sys
     assert sys.last_type is IndexError
     assert sys.last_value.args[0] == 'TEST'
     assert sys.last_traceback
+
+    # The next run should clear the exception info stored by the previous run
+    ItemMightRaise.raise_error = False
+    runner.pytest_runtest_call(ItemMightRaise())
+    assert sys.last_type is None
+    assert sys.last_value is None
+    assert sys.last_traceback is None
 
 
 def test_current_test_env_var(testdir, monkeypatch):

